crates/bevy_text/src/text2d.rs

```diff
@@ -125,42 +127,36 @@ pub fn extract_text2d_sprite(
 
 #[derive(Debug, Default)]
 pub struct QueuedText2d {
-    entities: Vec<Entity>,
+    entities: HashSet<Entity>,
 }
 
 /// Updates the layout and size information whenever the text or style is changed.
 /// This information is computed by the `TextPipeline` on insertion, then stored.
 #[allow(clippy::too_many_arguments, clippy::type_complexity)]
 pub fn text2d_system(
-    mut queued_text: Local<QueuedText2d>,
+    mut queue: Local<QueuedText2d>,
     mut textures: ResMut<Assets<Image>>,
     fonts: Res<Assets<Font>>,
     windows: Res<Windows>,
+    mut scale_factor_changed: EventReader<WindowScaleFactorChanged>,
     mut texture_atlases: ResMut<Assets<TextureAtlas>>,
     mut font_atlas_set_storage: ResMut<Assets<FontAtlasSet>>,
     mut text_pipeline: ResMut<DefaultTextPipeline>,
-    mut text_queries: ParamSet<(
-        Query<Entity, (With<Text2dSize>, Changed<Text>)>,
-        Query<(&Text, Option<&Text2dBounds>, &mut Text2dSize), With<Text2dSize>>,
+    mut text_query: Query<(
+        Entity,
+        Changed<Text>,
+        &Text,
+        Option<&Text2dBounds>,
+        &mut Text2dSize,
     )>,
 ) {
-    // Adds all entities where the text or the style has changed to the local queue
-    for entity in text_queries.p0().iter_mut() {
-        queued_text.entities.push(entity);
-    }
-
-    if queued_text.entities.is_empty() {
-        return;
-    }
-
+    // We need to consume the entire iterator, hence `last`
+    let factor_changed = scale_factor_changed.iter().last().is_some();
     let scale_factor = windows.scale_factor(WindowId::primary());
 
-    // Computes all text in the local queue
-    let mut new_queue = Vec::new();
-    let mut query = text_queries.p1();
-    for entity in queued_text.entities.drain(..) {
-        if let Ok((text, bounds, mut calculated_size)) = query.get_mut(entity) {
-            let text_bounds = match bounds {
+    for (entity, text_changed, text, maybe_bounds, mut calculated_size) in text_query.iter_mut() {
+        if factor_changed || text_changed || queue.entities.remove(&entity) {
```

Comment:
`queue.entities.remove` might look expensive, but the other changes more than drown out this overhead.